src/utils/asyncValidators.js

```diff
@@ -1,48 +1,83 @@
-import web3 from 'public-modules/Utilities/Web3Client';
-import config from 'public-modules/config';
+/* eslint no-throw-literal: 0 */
 
-const sleep = ms => new Promise(resolve => setTimeout(resolve, ms))
+import { actions } from 'layout/App/reducer';
+import { includes } from 'lodash';
 
-function* stubbed() {
-	yield sleep(2000).then(() => '1');
-	yield sleep(2000).then(() => '2');
-	return sleep(2000).then(() => '3');
-}
+const { getTokenBalance } = actions;
 
+const ZERO_BALANCE_ERROR = 'ZERO_BALANCE_ERROR';
+const INSUFFICIENT_BALANCE_ERROR = 'INSUFFICIENT_BALANCE_ERROR';
+const CONTRACT_DOES_NOT_CONFORM_TO_ERC20_ERROR =
+  'CONTRACT_DOES_NOT_CONFORM_TO_ERC20_ERROR';
+const UNKNOWN_ERROR = 'UNKNOWN_ERROR';
 
-const runPoll = () => new Promise(resolve => {
-	console.log('enter promise')
+const tokenValidation = (amount, tokenAddress, dispatch) => {
+  const handleResult = ([balance, symbol]) => {
+    balance = Number(balance);
+    amount = Number(amount);
 
-	const poll = generator => {
-		console.log('enter poll')
+    if (balance === 0) {
+      throw { error: ZERO_BALANCE_ERROR, balance, symbol };
+    } else if (balance < amount) {
+      throw { error: INSUFFICIENT_BALANCE_ERROR, balance, symbol };
+    }
+  };
 
-		if (!generator) {
-			generator = stubbed()
-		}
+  const handleRejection = e => {
+    if (includes('decode bytes32 from ABI', e.message)) {
+      throw { error: CONTRACT_DOES_NOT_CONFORM_TO_ERC20_ERROR };
+    }
 
-		const p = generator.next()
-		console.log(p)
+    throw { error: UNKNOWN_ERROR };
+  };
 
-		p.value.then(value => {
-			console.log(value);
-			if(!p.done) {
-				console.log('not done')
-				poll(generator)
-			} else {
-				resolve(value)
-			}
-		})
-	}
+  return new Promise((resolve, reject) => {
+    dispatch(getTokenBalance(tokenAddress, resolve, reject));
+  }).then(handleResult, handleRejection);
+};
 
-	poll()
-})
+const tokenValidationWrapper = (
+  values,
+  amountKey,
+  tokenContractKey,
+  dispatch
+) => {
+  if (!values[tokenContractKey]) return Promise.resolve();
 
-const tokenValidation = values => {
```

Comment:
:) so much better than the polling strategy

Reply:
:)